### PR TITLE
Add MySQL

### LINKED ISSUE #231: Spring Boot Database: MySQL

### PR DESCRIPTION
**Pull request by swarajsaaj**
Add support for MySQL database

Fix #231

I have replicated the postgresql style here without extracting common functionalities to avoid making a complicated if-else ladder in future.

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
Wow, thanks @swarajsaaj !
I'll review it soon

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    springPropertiesForTest(baseName).forEach((k, v) -> springBootPropertiesService.addPropertiesTest(project, k, v));
+  }
+
+  private Map<String, Object> springProperties(String baseName, String packageName) {
+    TreeMap<String, Object> result = new TreeMap<>();
+    result.put("spring.datasource.type", "com.zaxxer.hikari.HikariDataSource");
+    result.put("spring.datasource.url", "jdbc:mysql://localhost:3306/" + baseName);
+    result.put("spring.datasource.username", "root");
+    result.put("spring.datasource.password", "");
+    result.put("spring.datasource.driver-class-name", "com.mysql.jdbc.Driver");
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java:121]: I don't think we need this

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    result.put("spring.datasource.hikari.poolName", "Hikari");
+    result.put("spring.datasource.hikari.auto-commit", false);
+
+    result.put("spring.data.jpa.repositories.bootstrap-mode", "deferred");
+    result.put("spring.jpa.properties.hibernate.dialect", "org.hibernate.dialect.MySQL8Dialect");
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java:126]: same, I don't think we need this. What I can suggest, is to generate a JHipster project with MySQL and check all properties in application-prod.yml

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package tech.jhipster.lite.generator.server.springboot.database.mysql.domain;
+
+public class MySQL {
+
+  public static final String TESTCONTAINERS_VERSION = "1.16.0";
+
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java:6]: just an idea, maybe we can add: DOCKER_MYSQL_VERSION = "mysql:8.0.27"

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    result.put("spring.jpa.hibernate.ddl-auto", "none");
+    result.put("spring.jpa.hibernate.naming.physical-strategy", "org.springframework.boot.orm.jpa.hibernate.SpringPhysicalNamingStrategy");
+    result.put("spring.jpa.hibernate.naming.implicit-strategy", "org.springframework.boot.orm.jpa.hibernate.SpringImplicitNamingStrategy");
+    return result;
+  }
+
+  private Map<String, Object> springPropertiesForTest(String baseName) {
+    TreeMap<String, Object> result = new TreeMap<>();
+    result.put("spring.datasource.driver-class-name", "org.testcontainers.jdbc.ContainerDatabaseDriver");
+    result.put("spring.datasource.url", "jdbc:tc:mysql:8.0.27:///" + baseName);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java:147]: we could use MySQL.getDockerMysqlVersion(), following my comment above

File: src/main/resources/generator/server/springboot/database/mysql/mysql.yml.mustache [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# This configuration is intended for development purpose, it's **your** responsibility to harden it for production
+version: '3.8'
+services:
+  mysql:
+    image: mysql:8.0.27
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/database/mysql/mysql.yml.mustache:5]: we could use {{dockerMysqlVersion}}, injected by DOCKER_MYSQL_VERSION value

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  @Override
+  public void addHibernateCore(Project project) {
+    Dependency dependency = Dependency.builder().groupId("org.hibernate").artifactId("hibernate-core").build();
+
+    buildToolService.addDependency(project, dependency);
+  }
+
+  @Override
+  public void addDockerCompose(Project project) {
+    project.addDefaultConfig(BASE_NAME);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java:77]: we could add something like:

project.addConfig("dockerMysqlVersion", MySQL.getDockerMysqlVersion());

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  private MySQL() {}
+
+  public static String getTestcontainersVersion() {
+    return TESTCONTAINERS_VERSION;
+  }
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java:11]: with new method getDockerMysqlVersion()

File: src/test/java/tech/jhipster/lite/generator/server/springboot/database/mysql/application/MySQLApplicationServiceIT.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+    mysqlApplicationService.addTestContainers(project);
+
+    assertFileContent(project, POM_XML, "<testcontainers.version>");
+    assertFileContent(project, POM_XML, "</testcontainers.version>");
+    assertFileContent(project, POM_XML, testcontainers());
+    assertFileContent(
+      project,
+      getPath(TEST_RESOURCES, "config/application.properties"),
+      List.of("spring.datasource.url=jdbc:tc:mysql:8.0.27:///chips", "spring.datasource.username=chips")
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/server/springboot/database/mysql/application/MySQLApplicationServiceIT.java:194]: use MySQL.getDockerMysqlVersion()

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java (hunk 3 above)
Comment (swarajsaaj) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java:6]: yes sure, I think constant name could be DOCKER_IMAGE_NAME instead?

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java (hunk 2 above)
Comment (swarajsaaj) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java]: agreed, these properties are not needed. Removed

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java (hunk 1 above)
Comment (swarajsaaj) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQLDomainService.java]: agreed, these properties are not needed. Removed

File: src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java (hunk 3 above)
Comment (pascalgrimaud) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/database/mysql/domain/MySQL.java:6]: yes it's ok :)

**Comment by pascalgrimaud** [comment]
By reviewing this PR, I need to improve the code in PostgreSQL too :)

**Review by pascalgrimaud — APPROVED** [review]

**Comment by pascalgrimaud** [comment]
Don't forget to claim your bounty @swarajsaaj 
Well deserved

**Comment by swarajsaaj** [comment]
I claim the bounty :) https://opencollective.com/generator-jhipster/expenses/58753

**Comment by pascalgrimaud** [comment]
@swarajsaaj : approved!

**Pull request merged**